File: app/src/main/java/com/kazakago/cleanarchitecture/presentation/view/adapter/ForecastRecyclerAdapter.java
package com.kazakago.cleanarchitecture.presentation.view.adapter;

import android.content.Context;
import android.databinding.DataBindingUtil;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.kazakago.cleanarchitecture.BR;
import com.kazakago.cleanarchitecture.R;
import com.kazakago.cleanarchitecture.databinding.RecyclerForecastBinding;
import com.kazakago.cleanarchitecture.presentation.listener.view.adapter.ForecastRecyclerAdapterListener;
import com.kazakago.cleanarchitecture.presentation.presenter.adapter.ForecastViewModel;

import java.util.ArrayList;
import java.util.List;

/**
 * Forecast RecyclerView Adapter
 * <p>
 * Created by tamura_k on 2016/05/31.
 */
public class ForecastRecyclerAdapter extends RecyclerView.Adapter<ForecastRecyclerAdapter.ViewHolder> {

    static class ViewHolder extends RecyclerView.ViewHolder {

        private RecyclerForecastBinding binding;

        ViewHolder(View view) {
            super(view);
            binding = DataBindingUtil.bind(itemView);
        }

        public RecyclerForecastBinding getBinding() {
            return binding;
        }

    }

    private Context context;
    private List<ForecastViewModel> forecastViewModelList;
    private ForecastRecyclerAdapterListener forecastRecyclerAdapterListener;

    public ForecastRecyclerAdapter(Context context) {
        this.context = context;
        this.forecastViewModelList = new ArrayList<>();
    }

    @Override
    public int getItemCount() {
        return forecastViewModelList.size();
    }

    @Override
    public ForecastRecyclerAdapter.ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        View view = LayoutInflater.from(context).inflate(R.layout.recycler_forecast, parent, false);
        return new ForecastRecyclerAdapter.ViewHolder(view);
    }

    @Override
    public void onBindViewHolder(ForecastRecyclerAdapter.ViewHolder holder, int position) {
        final ForecastViewModel item = forecastViewModelList.get(position);

        holder.getBinding().setVariable(BR.viewModel, item);
        holder.getBinding().executePendingBindings();
        holder.getBinding().getRoot().setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                if (forecastRecyclerAdapterListener != null) forecastRecyclerAdapterListener.onItemClick(item);
            }
        });
    }

    public void setForecastViewModelList(List<ForecastViewModel> forecastViewModelList) {
        this.forecastViewModelList = forecastViewModelList;
    }

    public void setForecastRecyclerAdapterListener(ForecastRecyclerAdapterListener forecastRecyclerAdapterListener) {
        this.forecastRecyclerAdapterListener = forecastRecyclerAdapterListener;
    }

}

File: app/src/main/java/com/kazakago/cleanarchitecture/presentation/view/adapter/ForecastRecyclerAdapter.kt
package com.kazakago.cleanarchitecture.presentation.view.adapter

import android.content.Context
import android.support.v7.widget.RecyclerView
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.TextView
import com.kazakago.cleanarchitecture.R
import com.kazakago.cleanarchitecture.domain.model.weather.Forecast
import com.kazakago.cleanarchitecture.presentation.extension.formattedText
import com.squareup.picasso.Picasso

class ForecastRecyclerAdapter(private val context: Context) : RecyclerView.Adapter<ForecastRecyclerAdapter.ViewHolder>() {

    interface Listener {
        fun onItemClick(forecast: Forecast)
    }

    inner class ViewHolder(context: Context, parent: ViewGroup) : AbsViewHolder<Forecast>(context, parent, R.layout.recycler_forecast) {

        private val dateLabelTextView = itemView.findViewById<TextView>(R.id.dateLabelTextView)
        private val dateTextView = itemView.findViewById<TextView>(R.id.dateTextView)
        private val weatherImageView = itemView.findViewById<ImageView>(R.id.weatherImageView)
        private val telopTextView = itemView.findViewById<TextView>(R.id.telopTextView)
        private val maxTemperatureTextView = itemView.findViewById<TextView>(R.id.maxTemperatureTextView)
        private val minTemperatureTextView = itemView.findViewById<TextView>(R.id.minTemperatureTextView)

        override fun setItem(item: Forecast) {
            Picasso.with(context).load(item.imageUrl)
                    .into(weatherImageView)
            dateLabelTextView.text = item.dateLabel
            dateTextView.text = item.date.formattedText(context)
            telopTextView.text = item.telop
            maxTemperatureTextView.text = context.getString(R.string.temperature_max, item.maxTemperature?.toString() ?: "--")
            minTemperatureTextView.text = context.getString(R.string.temperature_min, item.minTemperature?.toString() ?: "--")
            itemView?.setOnClickListener {
                listener?.onItemClick(item)
            }
        }

    }

    var listener: Listener? = null
    var forecastList: List<Forecast> = listOf()

    override fun getItemCount(): Int = forecastList.size

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        return ViewHolder(context, parent)
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        holder.setItem(forecastList[position])
    }

}
